compiler/src/dotty/tools/dotc/printing/SyntaxHighlighting.scala

```diff
@@ -22,343 +29,108 @@ object SyntaxHighlighting {
-        } else {
-          newBuf += curr
-          closing = 0
-        }
+      object TreeHighlighter extends untpd.UntypedTreeTraverser {
+        import untpd._
 
-        if (Chars.isLineBreakChar(curr)) {
-          newBuf append LiteralColor
+        def ignored(tree: NameTree) = {
+          val name = tree.name.toTermName
+          // trees named <error> and <init> have weird positions
+          name == nme.ERROR || name == nme.CONSTRUCTOR
         }
-      }
-      newBuf append NoColor
-      prev = curr
-    }
-
-    def appendSingleQuote(delim: Char) = remaining.take(3) match {
-      case chr #:: '\'' #:: _ => // single character
-        newBuf append LiteralColor
-        newBuf appendAll s"'$chr'"
-        newBuf append NoColor
-        takeChars(2)
-        prev = '\''
-      case '\\' #:: chr #:: '\'' #:: _ => // escaped character
-        newBuf append LiteralColor
-        newBuf appendAll s"'\\$chr'"
-        newBuf append NoColor
-        takeChars(3)
-        prev = '\''
-      case _ => appendWhile(delim, !typeEnders.contains(_), literal)
-    }
 
-    def append(c: Char, shouldHL: String => Boolean, highlight: String => String) = {
-      var curr: Char = 0
-      val sb = new StringBuilder(s"$c")
-
-      def delim(c: Char) = (c: @switch) match {
-        case ' ' => true
-        case '\n' => true
-        case '(' => true
-        case ')' => true
-        case '[' => true
-        case ']' => true
-        case ':' => true
-        case '@' => true
-        case ',' => true
-        case '.' => true
-        case _ => false
-      }
-
-      val valDefStarterTokens = "var" :: "val" :: "def" :: "case" :: Nil
-
-      /** lastValDefToken is used to check whether we want to show something
-       *  in valDef color or not. There are only a few cases when lastValDefToken
-       *  should be updated, that way we can avoid stopping coloring too early.
-       *  eg.: case A(x, y, z) => ???
-       *  Without this function only x would be colored.
-       */
-      def updateLastToken(currentToken: String): String =
-        (lastValDefToken, currentToken) match {
-          case _ if valDefStarterTokens.contains(currentToken) => currentToken
-          case (("val" | "var"), "=") => currentToken
-          case ("case", ("=>" | "class" | "object")) => currentToken
-          case ("def", _) => currentToken
-          case _ => lastValDefToken
+        def highlight(trees: List[Tree])(implicit ctx: Context): Unit =
+          trees.foreach(traverse)
+
+        def traverse(tree: Tree)(implicit ctx: Context): Unit = {
+          tree match {
+            case tree: NameTree if ignored(tree) =>
+              ()
+            case tree: MemberDef /* ValOrDefDef | ModuleDef | TypeDef */ =>
+              for (annotation <- tree.rawMods.annotations)
+                highlightPosition(annotation.pos, AnnotationColor)
+              val color = if (tree.isInstanceOf[ValOrDefDef]) ValDefColor else TypeColor
```

Comment:
Use pattern match instead of `isInstanceOf`

Reply:
Maybe could just be another case in the outer pattern match and `for (annotation <- ...` could go in a helper method.